### ISSUE TITLE
Napari fails to create ellipse when specifying center and radius in 3 dimensions

### ISSUE THREAD

**Issue by bigtoasty** [napari/napari#4258]
## 🐛 Bug

The documentation for creating ellipses states that either the 4 corner vertices must be specified, or a center position and radius must be specified. I find that when attempting to create ellipses the second way, napari throws an error because it expects that the 4 corner method will be used instead. I suspect the issue originates in napari/layers/shapes/_shapes_models/ellipse.py on line 71 where it uses an if statement instead of else if. 

## To Reproduce

1. Attempt to add an ellipse to a layer in 3 dimensions by specifying a center and radius

Error message: 
```
~/anaconda3/lib/python3.7/site-packages/napari/layers/shapes/_shapes_models/ellipse.py in data(self, data)
     74                     "Data shape does not match a ellipse. Ellipse expects four corner vertices, {number} provided.",
     75                     deferred=True,
---> 76                     number=len(data),
     77                 )
     78             )
```

ValueError: Data shape does not match a ellipse. Ellipse expects four corner vertices, 2 provided.

Offending code in ellipse.py
```
@data.setter
    def data(self, data):
        data = np.array(data).astype(float)

        if len(self.dims_order) != data.shape[1]:
            self._dims_order = list(range(data.shape[1]))

        if len(data) == 2 and data.shape[1] == 2:
            data = center_radii_to_corners(data[0], data[1])

        if len(data) != 4:
            raise ValueError(
                trans._(
                    "Data shape does not match a ellipse. Ellipse expects four corner vertices, {number} provided.",
                    deferred=True,
                    number=len(data),
                )
            )

        self._data = data
```

**Comment by Carreau** [comment]
Hi there, 

I see that you are using Python 3.7 so you might not be getting the latest version of napari. 

Second I'm not sure ellipses can have arbitrary directions, I think they must be in in the x,y plane, i.e. they are not ellipsoids. 

>>> s.add_ellipses(np.array([[0,50],[50,30]]))

Works for me, but 

>>> s.add_ellipses(np.array([[0,0, 1],[50,30, 20]]))
Fail with similar error message, though I believe this
Is expected.

I also dont' think line 71 misses an else. 

The documentation state at some point " Internally we use the four vertex representation so we can always support rotated ellipses"

so the block on line 69 `data = center_radii_to_corners(data[0], data[1])` is made to transform the center/radii to 4 corners. After we expect `len(data)` to always be 4.

**Comment by Carreau** [comment]
That being said the code might have issues as `s.add_ellipses(np.array([[0,0],[0,100],[100,100], [2000, 2000]]))` works, and should likely fail as it is not a rectangle.

**Comment by bigtoasty** [comment]
The documentation doesn't specify that ellipses can't be added with center+radius in 3d so perhaps that should be added if that's really the expected behavior. 

"I also dont' think line 71 misses an else."

I wasn't paying too much attention and just thought it was weird that it checked to see if data was size 2 and then went to directly check if data was size 4. I missed that data gets changed within the first if block.

**Comment by Carreau** [comment]
> The documentation doesn't specify that ellipses can't be added with center+radius in 3d so perhaps that should be added if that's really the expected behavior.

Yeah, I'm unsure if it's an oversight in documentation or in the code.

"3D" is for many layer just a view where the "flat" layers are actually just projected Z=0 plane. So maybe 3D can actually be supported, the question is should it ? And then should it be a "shape" layer ? Or should we allow arbitrary 2D surfaces in any orientations ?

**Comment by 2dx** [comment]
I think center+radius in 3d creats 3d closed surface and not 2d shape list or projected one. Also specifying this way is reasonably denied by add_shape method. May should be accepted by add_surface method.

**Comment by jni** [comment]
@sofroniewn any thoughts here?

**Comment by sofroniewn** [comment]
We could let the ellipse accept a `2xD` array for center/ radii in more than two D. It should actually work almost out of the box as most of the functions I think would work in D dimensions, we just need to update a bunch of doc strings. The ellipse API will then accept either `2xD` for center radius or `4xD` for corner representation (though now I think about it that should maybe be `2**D x D` for the number of corners if all corners are specified .....

**Comment by sofroniewn** [comment]
@Carreau do you want to give implementing something based on https://github.com/napari/napari/issues/4258#issuecomment-1073413637 a go?

**Comment by Carreau** [comment]
I can give it a shot.

**Comment by Carreau** [comment]
> We could let the ellipse accept a `2xD` array for center/ radii in more than two D. It should actually work almost out of the box as most of the functions I think would work in D dimensions, we just need to update a bunch of doc strings. The ellipse API will then accept either `2xD` for center radius or `4xD` for corner representation (though now I think about it that should maybe be `2**D x D` for the number of corners if all corners are specified .....



> @Carreau do you want to give implementing something based on [#4258 (comment)](https://github.com/napari/napari/issues/4258#issuecomment-1073413637) a go?

For info, most of the 3D rendering of ellipses is completely borked, and render non planar aligned ellipses in wrong orientation and stretching. #4330 should start fixing the rendering.

So I've been banging my head as why other representation don't visually work. 



> `2**D x D` for the number of corners if all corners are specified

That should still be (4,D) for ellipses, (2**D, D) would be for D-ellipsoids but I don't think that should be in the shape layer, should it ? Do we want D-ellipsoids ? or D-shapes in General ?

https://miegakure.com/ and in particular https://www.youtube.com/watch?v=vZp0ETdD37E show some complexities with ND-shapes, and I'm unsure we want to tackle it in napari.

On top of that it would also be unclear to me how to manipulate a 2**D ellipsoid, in particular  if the multiple axes are not axis aligned.

**Comment by matham** [comment]
I am currently in need of something like a D-ellipsoids, I think. I have a bunch of cells which have different radii in each of the 3 dimensions (or at least in xy and z). And I think D-ellipsoids for D=3 is needed to do that. And each cell has a different radii so we can't just use dimension on the viewer to do that. 

The out-of-layer for points kinda does that, although I don't think we can set the marker size differently for xy/z.

BTW, the docs of `add_ellipses` was quite confusing there. Because I had thought it was able to do D-ellipsoids and that's what the `D` was referencing. So it was very confusing about why `4xD`. But I can see now that it just means it can do an ellipse that is rotated in `D` dimensions. Also, the order of the 4-corners (clockwise/counterclockwise) is not specified. So it may help to clarify the docs.

**Comment by Czaki** [comment]
@matham If I correctly understand you need, you need to use surface layers for rendering 3D shape. 
Unfortunately, we do not have a function to generate surface for 3D elipsoid as far as I know.

**Comment by matham** [comment]
I'm not sure I want surfaces because I use them to highlight cells in the images. Shapes and points are perfect for that because it can mark the boundary of the cell in the current layer without painting over the actual image. Surfaces I think highlight the volume, kinda obscuring it? Also, we have millions of cells in the 3d data so I don't expect surfaces would be able to handle that!?

**Comment by Czaki** [comment]
So you want a single ellipse per layer?

**Comment by matham** [comment]
Yes. An ellipse that is shown in each z layer, whose xy size at a particular z slice scales with the elipsoid's z radius and distance from ellipsez  center. Kind of like if the points layer had a size value for x, y and z instead of just one value for all 3 dims so the radius is the same in all 3 axes.

**Comment by TimMonko** [comment]
@matham. Can you describe your data more? Are the ellipses biological in origin or a result of the imaging? It sounds like the data you are describing might be anisotropic.

Here's an explanation why points (and perhaps others) are isotoropic: https://github.com/napari/napari/pull/5582
You can change the visual representation of data so that it's isotropic with the `scale` property, such as `viewer.layers[0].scale = (5,1,1)`
And here's a way to scale data so that the underlying data itself is isotropic: https://haesleinhuepf.github.io/BioImageAnalysisNotebooks/19_spatial_transforms/scaling.html

**Comment by matham** [comment]
Thanks for clarifying this. Our data is indeed anisotropic because it's light sheet data with z and xy have different resolution.

I don't love rescaling our data since it's not worth doing just for visualizing. But, I can just use the average diameter of z and xy together with layer scaling to be good enough.

It would be nice though if the docs make it clear in the points layer that it doesn't support anisotropic points and similarly for shapes with respect to 3-D shapes. Just so that others don't get sucked into trying to figure out how to do it when it's not supported.

**Comment by TimMonko** [comment]
> It would be nice though if the docs make it clear in the points layer that it doesn't support anisotropic points and similarly for shapes with respect to 3-D shapes. Just so that others don't get sucked into trying to figure out how to do it when it's not supported.

I've opened an issue in the docs repo https://github.com/napari/docs/issues/740, though perhaps we can also update docstrings. If you'd like to contribute this update, we'd be very grateful! If you'd like to, and want some guidance I'd be happy to help :)
